packages/common/container-definitions/src/loader.ts

```diff
@@ -102,23 +102,147 @@ export interface ICodeAllowList {
+     * - {@link IContainer.connect}
+     */
     (event: "connected", listener: (clientId: string) => void);
+
+    /**
+     * Fires when new container code details have been proposed, prior to acceptance.
+     *
+     * @remarks Listener parameters:
+     *
+     * - `codeDetails`: The code details being proposed.
+     *
+     * - `proposal`: NOT RECOMMENDED FOR USE.
+     *
+     * @see {@link IContainer.proposeCodeDetails}
+     */
     (event: "codeDetailsProposed", listener: (codeDetails: IFluidCodeDetails, proposal: ISequencedProposal) => void);
+
+    /**
+     * @deprecated No replacement API recommended.
+     */
     (event: "contextChanged", listener: (codeDetails: IFluidCodeDetails) => void);
-    (event: "disconnected" | "attached", listener: () => void);
+
+    /**
+     * Emitted when the {@link IContainer} becomes disconnected from the Fluid service.
+     *
+     * @remarks Reflects connection state changes against the (delta) service acknowledging ops/edits.
+     *
+     * @see
+     *
+     * - {@link IContainer.connectionState}
+     *
+     * - {@link IContainer.disconnect}
+     */
+    (event: "disconnected", listener: () => void);
+
+    /**
+     * Emitted when a {@link AttachState.Detached | detached} container is
+     * {@link AttachState.Attached | attached} to the Fluid service.
+     *
+     * @see
+     *
+     * - {@link IContainer.attachState}
+     *
+     * - {@link IContainer.attach}
+     */
+    (event: "attached", listener: () => void);
+
+    /**
+     * Emitted when the {@link IContainer} is closed, which permanently disables it.
+     *
+     * @remarks Listener parameters:
+     *
+     * - `error`: If the container was closed due to error, this will contain details about the error that caused it.
+     *
+     * @see {@link IContainer.close}
+     */
     (event: "closed", listener: (error?: ICriticalContainerError) => void);
+
+    /**
+     * Emitted when the container encounters a state which may lead to errors, which may be actionable by the consumer.
+     *
+     * @remarks
+     *
+     * Note: this event is not intended for general use.
+     * The longer-term intention is to surface warnings more directly on the APIs that produce them.
+     * For now, use of this should be avoided when possible.
+     *
+     * Listener parameters:
+     *
+     * - `error`: The warning describing the encountered state.
+     */
     (event: "warning", listener: (error: ContainerWarning) => void);
+
+    /**
+     * Emitted immediately after processing an incoming operation (op).
+     *
+     * @remarks
+     *
+     * Note: this event is not intended for general use.
+     * Prefer to listen to events on the appropriate ultimate recipients of the ops, rather than listening to the
+     * ops directly on the {@link IContainer}.
+     *
+     * Listener parameters:
+     *
+     * - `message`: The op that was processed.
+     */
     (event: "op", listener: (message: ISequencedDocumentMessage) => void);
```

Comment:
we should deprecated this. consumers shouldn't be spying the op stream, they can do a lot of harm here

Reply:
We do have usages of this in the repo. What would the alternative be?

Reply:
I'm going to leave this be for now. This probably merits further discussion and its own PR.